napari/utils/progress.py

```diff
@@ -0,0 +1,245 @@
+        total: Optional[int] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+
+        # check if there's a napari viewer instance
+        viewer = get_viewer_instance()
+        self.has_viewer = viewer is not None
+        if self.has_viewer:
+            kwargs['gui'] = True
+
+        kwargs = kwargs.copy()
+        pbar_kwargs = {k: kwargs.pop(k) for k in set(kwargs) - _tqdm_kwargs}
+
+        self.nested_token = None
+
+        super().__init__(iterable, desc, total, *args, **kwargs)
+        if not self.has_viewer:
+            return
+
+        self._pbar = get_pbar(viewer, **pbar_kwargs)
+        if self.total is not None:
+            self._pbar.setRange(self.n, self.total)
+            self._pbar._set_value(self.n)
+        else:
+            self._pbar.setRange(0, 0)
+
+        if desc:
+            self.set_description(desc)
+        else:
+            desc = get_calling_function_name(max_depth=5)
+            if desc:
+                self.set_description(desc)
+            else:
+                # TODO: pick a better default
+                self.set_description("Progress Bar")
+
+        self.show()
+        QApplication.processEvents()
+
+    def __enter__(self):
+        self.nested_token = IS_NESTED.set(True)  # noqa
+        return super().__enter__()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if IS_NESTED.get():
+            IS_NESTED.reset(self.nested_token)
+        if self.has_viewer:
+            self._pbar.parentWidget().close()
+        return super().__exit__(exc_type, exc_value, traceback)
+
+    def display(self, msg: str = None, pos: int = None) -> None:
+        """Update the display."""
+        if not self.has_viewer:
+            return super().display(msg=msg, pos=pos)
+
+        if self.total != 0:
+            etas = str(self).split('|')[-1]
+            self._pbar._set_value(self.n)
+            self._pbar._set_eta(etas)
+        QApplication.processEvents()
+
+    def set_description(self, desc):
+        """Update progress bar description"""
+        super().set_description(desc, refresh=True)
+        if self.has_viewer:
+            self._pbar._set_description(self.desc)
+
+    def hide(self):
+        """Hide the progress bar"""
+        if self.has_viewer:
+            self._pbar.hide()
+
+    def show(self):
+        """Show the progress bar"""
+        if self.has_viewer:
+            self._pbar.show()
+
+    def close(self):
+        """Closes and deletes the progress bar widget"""
+        if self.disable:
+            return
+        if self.has_viewer:
+            # still need to close groups of pbars outside context
+            if not IS_NESTED.get():
+                self._pbar.parentWidget().close()
+            # otherwise we just close the current progress bar
+            self._pbar.close()
+        super().close()
+
+
+def progrange(*args, **kwargs):
+    return progress(range(*args), **kwargs)
+
+
+class ProgressBar(QWidget):
```

Comment:
this needs to move into `_qt` and `progress` should "fail gracefully" if we're running headless (i.e. even if `qtpy` isn't availabled)

Reply:
ok @tlambert03 so I've done this now by moving every mention of `qt` out of `progress` except for importing `get_bar` from `_qt.widgets.qt_progress_bar` which I've put in the `__init__` in a `try-except` block. Is that the right approach?